File: solutions/ruby/palindrome-products/1/palindrome_products.rb
class Palindromes
  def initialize(max_factor:, min_factor: 1)
    raise ArgumentError, "min must be <= max" if min_factor > max_factor
    @max_factor = max_factor
    @min_factor = min_factor
  end

  def generate
    @palindromes = []
    (@min_factor..@max_factor).each do |i|
      (@min_factor..@max_factor).each do |j|
        product = i * j
        if palindrome?(product)
          if @palindromes.map(&:value).include?(product)
            existing = @palindromes.find { |p| p.value == product }
            existing.factors << [i, j] if (existing.factors.include?([i, j]) == false) and (existing.factors.include?([j, i]) == false)
          else
            @palindromes << PalindromeProduct.new(product, [[i, j]]) if palindrome?(product)
          end
        end
      end
    end
  end

  def smallest
    @palindromes.min_by(&:value) || PalindromeProduct.new(nil, [])
  end

  def largest
    @palindromes.max_by(&:value) || PalindromeProduct.new(nil, [])
  end
  
  def palindrome?(number)
    str = number.to_s
    str == str.reverse
  end
end

PalindromeProduct = Struct.new(:value, :factors)

File: solutions/csharp/palindrome-products/1/PalindromeProducts.cs
using System.ComponentModel;

public static class PalindromeProducts
{
    public static (int, IEnumerable<(int,int)>) Largest(int minFactor, int maxFactor)
    {
        var palindromeProducts = new Dictionary<int, IEnumerable<(int, int)>>();
        for (var i = minFactor; i <= maxFactor; i++)
        {
            for (var j = i; j <= maxFactor; j++)
            {
                var product = (i * j).ToString();
                var digits = product.ToCharArray();
                Array.Reverse(digits);
                if (new string(digits) == product)
                {
                    if (palindromeProducts.ContainsKey(i * j))
                    {
                        var currentFactors = palindromeProducts[i * j];
                        currentFactors = currentFactors.Append((i, j));
                        palindromeProducts[i * j] = currentFactors;
                    }
                    else
                    {
                        palindromeProducts[i * j] = [(i, j)];
                    }
                }
            }
        }

        if (palindromeProducts.Count == 0)
        {
            throw new ArgumentException();
        }

        var largest = palindromeProducts.Keys.Max();
        return (largest, palindromeProducts[largest]);
    }

    public static (int, IEnumerable<(int, int)>) Smallest(int minFactor, int maxFactor)
    {
        var palindromeProducts = new Dictionary<int, IEnumerable<(int, int)>>();
        for (var i = minFactor; i <= maxFactor; i++)
        {
            for (var j = i; j <= maxFactor; j++)
            {
                var product = (i * j).ToString();
                var digits = product.ToCharArray();
                Array.Reverse(digits);
                if (new string(digits) == product)
                {
                    if (palindromeProducts.ContainsKey(i * j))
                    {
                        var currentFactors = palindromeProducts[i * j];
                        currentFactors = currentFactors.Append((i, j));
                        palindromeProducts[i * j] = currentFactors;
                    }
                    else
                    {
                        palindromeProducts[i * j] = [(i, j)];
                    }
                }
            }
        }

        if (palindromeProducts.Count == 0)
        {
            throw new ArgumentException();
        }
        
        var smallest = palindromeProducts.Keys.Min();
        return (smallest, palindromeProducts[smallest]);
    }
}
